File: app/src/main/java/forpdateam/ru/forpda/views/messagepanel/attachments/AttachmentsPopup.java
package forpdateam.ru.forpda.views.messagepanel.attachments;

import android.content.Context;
import android.support.design.widget.BottomSheetDialog;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.api.RequestFile;
import forpdateam.ru.forpda.api.theme.editpost.models.AttachmentItem;
import forpdateam.ru.forpda.api.theme.editpost.models.EditPostForm;
import forpdateam.ru.forpda.views.messagepanel.AutoFitRecyclerView;
import forpdateam.ru.forpda.views.messagepanel.MessagePanel;

/**
 * Created by radiationx on 09.01.17.
 */

public class AttachmentsPopup {
    private BottomSheetDialog dialog;
    private MessagePanel messagePanel;
    private View bottomSheet;
    private AutoFitRecyclerView recyclerView;
    private AttachmentAdapter adapter = new AttachmentAdapter();

    private TextView noAttachments;
    private RelativeLayout textControls;
    private ImageButton addFile, deleteFile;
    private Button addToSpoiler, addToText;
    private FrameLayout progressOverlay;

    private OnInsertAttachmentListener insertAttachmentListener;

    public AttachmentsPopup(Context context, MessagePanel panel) {
        messagePanel = panel;
        dialog = new BottomSheetDialog(context);
        //dialog.setPeekHeight(App.getKeyboardHeight());
        //dialog.getWindow().getDecorView().setFitsSystemWindows(true);

        bottomSheet = View.inflate(context, R.layout.message_panel_attachments, null);
        recyclerView = (AutoFitRecyclerView) bottomSheet.findViewById(R.id.auto_fit_recycler_view);
        progressOverlay = (FrameLayout) bottomSheet.findViewById(R.id.progress_overlay);

        noAttachments = (TextView) bottomSheet.findViewById(R.id.no_attachments_text);
        textControls = (RelativeLayout) bottomSheet.findViewById(R.id.text_controls);
        addFile = (ImageButton) bottomSheet.findViewById(R.id.add_file);
        deleteFile = (ImageButton) bottomSheet.findViewById(R.id.delete_file);
        addToSpoiler = (Button) bottomSheet.findViewById(R.id.add_to_spoiler);
        addToText = (Button) bottomSheet.findViewById(R.id.add_to_text);

        recyclerView.setColumnWidth(App.getInstance().dpToPx(112));
        recyclerView.setAdapter(adapter);

        /*addFile.setOnClickListener(v -> {
            uploadFiles();
        });*/
        //deleteFile.setOnClickListener(v -> adapter.deleteSelected());

        adapter.setOnSelectedListener(this::onSelected);
        adapter.setOnDataChangeListener(this::onDataChange);
        /*adapter.setReloadOnClickListener(item -> {
            if (item.getLoadState() == AttachmentItem.STATE_NOT_LOADED) {
                item.setLoadState(AttachmentItem.STATE_LOADING);
                adapter.notifyItemLoadResult(item);
                new Handler().postDelayed(() -> {
                    item.setLoadState(AttachmentItem.STATE_LOADED);
                    adapter.notifyItemLoadResult(item);
                }, 3000);
            }
        });*/
        onDataChange(0);

        addToText.setOnClickListener(v -> insertAttachment(adapter.getSelected(), false));
        addToSpoiler.setOnClickListener(v -> insertAttachment(adapter.getSelected(), true));

        messagePanel.addAttachmentsOnClickListener(v -> {
            if (bottomSheet != null && bottomSheet.getParent() != null && bottomSheet.getParent() instanceof ViewGroup) {
                ((ViewGroup) bottomSheet.getParent()).removeView(bottomSheet);
            }
            dialog.setContentView(bottomSheet);
            dialog.show();
        });

        /*if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            Window window = dialog.getWindow();
            if (window != null) {
                window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
                }
            }
        }*/
    }

    public void insertAttachment(List<AttachmentItem> items, boolean toSpoiler) {
        StringBuilder text = new StringBuilder();
        if (toSpoiler)
            text.append("[spoiler]");
        for (AttachmentItem item : items) {
            if (insertAttachmentListener != null) {
                text.append(insertAttachmentListener.onInsert(item));
            } else {
                text.append("[attachment=").append(item.getId()).append(":").append(item.getName()).append("]");
            }
        }
        if (toSpoiler)
            text.append("[/spoiler]");
        messagePanel.insertText(text.toString());
        adapter.unSelectItems();
    }

    private void onDataChange(int count) {
        if (count > 0) {
            noAttachments.setText("Вложений: " + count);
            //dialog.setPeekHeight(App.getKeyboardHeight());
        } else {
            noAttachments.setText("Нет вложений");
            //dialog.setPeekHeight(App.px48);
        }
    }

    private void onSelected(AttachmentItem item, int index, int selected) {
        int firstGroup = selected > 0 ? View.GONE : View.VISIBLE;
        int secondGroup = selected > 0 ? View.VISIBLE : View.GONE;

        if (noAttachments.getVisibility() != firstGroup)
            noAttachments.setVisibility(firstGroup);
        if (addFile.getVisibility() != firstGroup)
            addFile.setVisibility(firstGroup);
        if (textControls.getVisibility() != secondGroup)
            textControls.setVisibility(secondGroup);
        if (deleteFile.getVisibility() != secondGroup)
            deleteFile.setVisibility(secondGroup);

        tryLockControls(!adapter.containNotLoaded());
    }

    private void tryLockControls(boolean enable) {
        if (textControls.getVisibility() == View.VISIBLE) {
            addToSpoiler.setEnabled(enable);
            addToText.setEnabled(enable);
            deleteFile.setEnabled(enable);
        }
    }

    public void setAddOnClickListener(View.OnClickListener listener) {
        addFile.setOnClickListener(listener);
    }

    public void setDeleteOnClickListener(View.OnClickListener listener) {
        deleteFile.setOnClickListener(listener);
    }

    public void onLoadAttachments(EditPostForm form) {
        adapter.add(form.getAttachments());
    }

    public List<AttachmentItem> preUploadFiles(List<RequestFile> files) {
        List<AttachmentItem> loadingItems = new ArrayList<>();
        for (RequestFile file : files) {
            AttachmentItem item = new AttachmentItem(file.getFileName());
            item.setProgressListener(percent -> {

            });
            Log.e("FORPDA_LOG", "ADD LOADING ITEM " + item);
            adapter.add(item);
            loadingItems.add(item);
        }
        return loadingItems;
    }

    public void onUploadFiles(List<AttachmentItem> items) {
        for(AttachmentItem item: items){
            Log.d("SUKA", "LOADED ITEM "+item.getName());
        }
        for (AttachmentItem item : items) {
            Log.e("FORPDA_LOG", "LOADING ITEM " + " : " + item);
            if (item.getLoadState() == AttachmentItem.STATE_NOT_LOADED) {
                adapter.removeItem(item);
                //SHOW ERROR
            } else {
                adapter.notifyItemLoadResult(item);
            }
        }
    }

    public void preDeleteFiles() {
        //block ui
        progressOverlay.setVisibility(View.VISIBLE);
        tryLockControls(false);
    }

    public void setAttachments(Collection<AttachmentItem> items) {
        adapter.clear();
        adapter.add(items);
    }

    public void clearAttachments() {
        adapter.clear();
    }


    public void onDeleteFiles(List<AttachmentItem> deletedItems) {
        //unblock ui
        Log.e("FORPDA_LOG", "ON DELETE FILES "+deletedItems);
        for (AttachmentItem item : deletedItems) {
            Log.e("FORPDA_LOG", "DELETED FILE "+item);
            messagePanel.setText(messagePanel.getMessage().replaceAll("\\[attachment=['\"]?" + item.getId() + ":[^\\]]*?]",""));
        }
        progressOverlay.setVisibility(View.GONE);
        adapter.deleteSelected();
    }

    public List<AttachmentItem> getAttachments() {
        return adapter.getItems();
    }

    public List<AttachmentItem> getSelected() {
        return adapter.getSelected();
    }

    /*private AttachmentItem getItemByName(String name) {
        for (AttachmentItem item : loadingItems) {
            if (item.getName().equals(name)) return item;
        }
        return null;
    }*/

    public void setInsertAttachmentListener(OnInsertAttachmentListener insertAttachmentListener) {
        this.insertAttachmentListener = insertAttachmentListener;
    }

    public interface OnInsertAttachmentListener {
        String onInsert(AttachmentItem item);
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/views/messagepanel/attachments/AttachmentsPopup.kt
package forpdateam.ru.forpda.ui.views.messagepanel.attachments

import android.content.Context
import com.google.android.material.bottomsheet.BottomSheetDialog
import androidx.recyclerview.widget.GridLayoutManager
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageButton
import android.widget.RelativeLayout
import android.widget.TextView

import java.util.ArrayList

import forpdateam.ru.forpda.App
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.entity.remote.editpost.AttachmentItem
import forpdateam.ru.forpda.entity.remote.editpost.EditPostForm
import forpdateam.ru.forpda.model.data.remote.api.RequestFile
import forpdateam.ru.forpda.ui.views.messagepanel.AutoFitRecyclerView
import forpdateam.ru.forpda.ui.views.messagepanel.MessagePanel

/**
 * Created by radiationx on 09.01.17.
 */

class AttachmentsPopup(context: Context, private val messagePanel: MessagePanel) {
    private val dialog: BottomSheetDialog
    private val bottomSheet: View?
    private val recyclerView: AutoFitRecyclerView
    private val adapter = AttachmentAdapter()

    private val noAttachments: TextView
    private val textControls: RelativeLayout
    private val addFile: ImageButton
    private val deleteFile: ImageButton
    private val addToSpoiler: Button
    private val addToText: Button
    private val progressOverlay: FrameLayout

    private var enabledTextControls = true
    private var isLinear = true
    private var isReverse = false


    private val attachments = mutableListOf<AttachmentItem>()
    private val selected = mutableListOf<AttachmentItem>()


    private var insertAttachmentListener: OnInsertAttachmentListener? = null

    fun getAttachments(): List<AttachmentItem> = attachments

    fun getSelected(): List<AttachmentItem> = selected

    init {
        dialog = BottomSheetDialog(context)
        val lp = dialog.window!!.attributes
        lp.dimAmount = 1.0f
        dialog.window!!.attributes = lp
        dialog.window!!.addFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND)
        //dialog.setPeekHeight(App.getKeyboardHeight());
        //dialog.getWindow().getDecorView().setFitsSystemWindows(true);

        bottomSheet = View.inflate(context, R.layout.message_panel_attachments, null)
        recyclerView = bottomSheet!!.findViewById<View>(R.id.auto_fit_recycler_view) as AutoFitRecyclerView
        progressOverlay = bottomSheet.findViewById<View>(R.id.progress_overlay) as FrameLayout

        noAttachments = bottomSheet.findViewById<View>(R.id.no_attachments_text) as TextView
        textControls = bottomSheet.findViewById<View>(R.id.text_controls) as RelativeLayout
        addFile = bottomSheet.findViewById<View>(R.id.add_file) as ImageButton
        deleteFile = bottomSheet.findViewById<View>(R.id.delete_file) as ImageButton
        addToSpoiler = bottomSheet.findViewById<View>(R.id.add_to_spoiler) as Button
        addToText = bottomSheet.findViewById<View>(R.id.add_to_text) as Button

        recyclerView.setColumnWidth(App.get().dpToPx(112, recyclerView.context))
        adapter.updateIsLinear(isLinear)
        adapter.updateReverse(isReverse)
        recyclerView.setFakeLinear(isLinear)
        recyclerView.adapter = adapter

        recyclerView.manager.spanSizeLookup = object : androidx.recyclerview.widget.GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(i: Int): Int {
                return if (isLinear) {
                    1
                } else if (i == 0) {
                    recyclerView.manager.spanCount
                } else {
                    1
                }
            }
        }

        /*addFile.setItemClickListener(v -> {
            uploadFiles();
        });*/
        //deleteFile.setItemClickListener(v -> adapter.deleteSelected());
        adapter.setReloadOnClickListener(object : AttachmentAdapter.OnReloadClickListener {
            override fun onReloadClick(item: AttachmentItem) {

            }
        })

        adapter.setOnItemClickListener(object : AttachmentAdapter.OnItemClickListener {
            override fun onItemClick(item: AttachmentItem) {
                item.toggle()
                if (item.isSelected) {
                    if (!selected.contains(item)) {
                        selected.add(item)
                    }
                } else {
                    selected.remove(item)
                }
                onSelectedChange()
                adapter.updateItem(item)
            }
        })
        adapter.setSelectorListener(object : AttachmentAdapter.SelectorListener {
            override fun onViewTypeChanged(isLinear: Boolean) {
                this@AttachmentsPopup.isLinear = isLinear
                recyclerView.setFakeLinear(isLinear)
                adapter.updateIsLinear(isLinear)
            }

            override fun onReverseClick() {
                isReverse = !isReverse
                adapter.updateReverse(isReverse)
                adapter.clear()
                adapter.add(attachments)
            }
        })
        onDataChange(0)

        addToText.setOnClickListener { v -> insertAttachment(selected, false) }
        addToSpoiler.setOnClickListener { v -> insertAttachment(selected, true) }

        messagePanel.addAttachmentsOnClickListener { v ->
            if (bottomSheet != null && bottomSheet.parent != null && bottomSheet.parent is ViewGroup) {
                (bottomSheet.parent as ViewGroup).removeView(bottomSheet)
            }
            dialog.setContentView(bottomSheet)
            dialog.show()
        }

        /*if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            Window window = dialog.getWindow();
            if (window != null) {
                window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
                }
            }
        }*/
    }

    fun setEnabledTextControls(enabled: Boolean) {
        enabledTextControls = enabled
    }

    fun insertAttachment(items: List<AttachmentItem>, toSpoiler: Boolean) {
        val text = StringBuilder()
        if (toSpoiler)
            text.append("[spoiler]")
        for (item in items) {
            if (insertAttachmentListener != null) {
                text.append(insertAttachmentListener!!.onInsert(item))
            } else {
                text.append("[attachment=").append(item.id).append(":").append(item.name).append("]")
            }
        }
        if (toSpoiler)
            text.append("[/spoiler]")
        messagePanel.insertText(text.toString())
        unSelectItems()
        dialog.cancel()
    }

    fun unSelectItems() {
        for (item in selected) {
            if (item.isSelected) item.toggle()
            adapter.updateItem(item)
        }
        selected.clear()
        onSelectedChange()
    }

    fun containNotLoaded(): Boolean {
        for (item in selected) {
            if (item.loadState != AttachmentItem.STATE_LOADED)
                return true
        }
        return false
    }


    fun deleteSelected() {
        for (item in selected) {
            if (item.status == AttachmentItem.STATUS_REMOVED) {
                attachments.remove(item)
                adapter.removeItem(item)
                updateDataCounter()
            }
        }
        unSelectItems()
    }


    private fun onDataChange(count: Int) {
        messagePanel.updateAttachmentsCounter(count)
        if (count > 0) {
            noAttachments.text = String.format(App.get().getString(R.string.attachments_count), count)
            //dialog.setPeekHeight(App.getKeyboardHeight());
        } else {
            noAttachments.setText(R.string.no_attachments)
            //dialog.setPeekHeight(App.px48);
        }
    }

    private fun updateDataCounter() {
        onDataChange(attachments.size)
    }

    private fun onSelectedChange() {
        val firstGroup = if (selected.size > 0) View.GONE else View.VISIBLE
        val secondGroup = if (selected.size > 0) View.VISIBLE else View.GONE

        if (!enabledTextControls) {
            noAttachments.visibility = View.VISIBLE
        } else if (noAttachments.visibility != firstGroup) {
            noAttachments.visibility = firstGroup
        }
        if (addFile.visibility != firstGroup)
            addFile.visibility = firstGroup
        if (!enabledTextControls) {
            textControls.visibility = View.GONE
        } else if (textControls.visibility != secondGroup) {
            textControls.visibility = secondGroup
        }
        if (deleteFile.visibility != secondGroup)
            deleteFile.visibility = secondGroup

        tryLockControls(!containNotLoaded())
    }

    private fun tryLockControls(enable: Boolean) {
        if (textControls.visibility == View.VISIBLE) {
            addToSpoiler.isEnabled = enable
            addToText.isEnabled = enable
            deleteFile.isEnabled = enable
        }
    }


    fun setAddOnClickListener(listener: () -> Unit) {
        addFile.setOnClickListener { listener.invoke() }
    }

    fun setDeleteOnClickListener(listener: () -> Unit) {
        deleteFile.setOnClickListener { listener.invoke() }
    }

    fun onLoadAttachments(form: EditPostForm) {
        attachments.addAll(form.attachments)
        adapter.add(form.attachments)
        updateDataCounter()
    }

    fun preUploadFiles(files: List<RequestFile>): List<AttachmentItem> {
        Log.d(LOG_TAG, "preUploadFiles $files")
        val loadingItems = ArrayList<AttachmentItem>()
        for (file in files) {
            val item = AttachmentItem(file.fileName)
            item.setProgressListener { percent ->

            }
            Log.d(LOG_TAG, "Add loading item $item")
            attachments.add(item)
            adapter.add(item)
            loadingItems.add(item)
        }
        return loadingItems
    }

    fun onUploadFiles(items: List<AttachmentItem>) {
        Log.d(LOG_TAG, "onUploadFiles $items")
        for (item in items) {
            Log.d(LOG_TAG, "Loading item $item")
            if (item.loadState == AttachmentItem.STATE_NOT_LOADED) {
                attachments.remove(item)
                selected.remove(item)
                adapter.removeItem(item)
                //SHOW ERROR
            } else {
                adapter.updateItem(item)
            }
        }
        updateDataCounter()
        onSelectedChange()
    }

    fun preDeleteFiles() {
        //block ui
        progressOverlay.visibility = View.VISIBLE
        tryLockControls(false)
    }

    fun setAttachments(items: List<AttachmentItem>) {
        clearAttachments()
        attachments.addAll(items)
        adapter.add(items)
        updateDataCounter()
    }

    fun clearAttachments() {
        attachments.clear()
        selected.clear()
        adapter.clear()
        updateDataCounter()
        onSelectedChange()
    }


    fun onDeleteFiles(deletedItems: List<AttachmentItem>) {
        //unblock ui
        Log.d(LOG_TAG, "onDeleteFiles $deletedItems")
        for (item in deletedItems) {
            Log.d(LOG_TAG, "Delete file $item")
            messagePanel.setText(messagePanel.message.replace(("\\[attachment=['\"]?" + item.id + ":[^\\]]*?]").toRegex(), ""))
        }
        progressOverlay.visibility = View.GONE
        deleteSelected()
    }

    fun setInsertAttachmentListener(insertAttachmentListener: OnInsertAttachmentListener) {
        this.insertAttachmentListener = insertAttachmentListener
    }

    interface OnInsertAttachmentListener {
        fun onInsert(item: AttachmentItem): String
    }

    companion object {
        private val LOG_TAG = AttachmentsPopup::class.java.simpleName
    }
}
